### DISCUSSION TITLE
Dependencies and "global" or multi-file variable.

### DISCUSSION THREAD

**Discussion by DenuxPlays** [Orange-OpenSource/hurl#2824] [Ideas]
Hey,
I am currently running into an issue that a few tests (but not all!) needs some extra preparation (in my case creating 2 accounts and saving their id).
I have to copy the requests captures assertions etc. to each file where its needed.
Something like this would be great:
Test that needs preparation:
```hurl
DEPENDS create-account.hurl
[Captures]
acc1_id: filevar acc1_id
acc2_id: filevar acc2_id
```
create-account.hurl
```hurl
POST https://api.test/account
[Captures]
acc1_id: jsonpath "$['id']"


POST https://api.test/account
[Captures]
acc2_id: jsonpath "$['id']"

export acc1_id
export acc2_id
```

Maybe something like this could be made possible.
The example was just out of nowhere so maybe someone has a better idea on how this could work but something like this is really missing in hurl.
(But not as important as doing basic math, IDE/text editors plugins etc.)
Maybe someone else can share their thoughts about this here.

**Comment by jcamiel** [comment]
Hi @DenuxPlays 

This issue is really about how to import Hurl fragment into another Hurl files. There are multiple issues and we know it's a feature important for some users:

Other related issue are #115 #1273 #317 #2110.

For the moment, mainly due to implementation complexity, we aren't going to develop "import" files soon. We need to take time to find a good syntax and also to deal with how to report error from a "stack" of included files for instance.

Hope it's not too harsh from us!

**Reply by DenuxPlays** [reply]
> Hi @DenuxPlays 
> 
> 
> 
> This issue is really about how to import Hurl fragment into another Hurl files. There are multiple issues and we know it's a feature important for some users:
> 
> 
> 
> Other related issue are #115 #1273 #317 #2110.
> 
> 
> 
> For the moment, mainly due to implementation complexity, we aren't going to develop "import" files soon. We need to take time to find a good syntax and also to deal with how to report error from a "stack" of included files for instance.
> 
> 
> 
> Hope it's not too harsh from us!

Hey,

No that is totally understandable and I would prefer a good syntax over a faster implementation.
Maybe I could help an drop some ideas on how to do that.
I do not make a promise here as I don't know when I can find time for this but if I can I will drop some ideas.
Maybe if we can agree on a syntax than a community member (or perhaps me) can implement this feature.

For now.
A workaround that I use for my login stuff is the following:
For login you create a session with your credentials that returns a token.
We do that by simply executing a curl request and setting the returned token inside an env that can be used by hurl.

**Reply by DenuxPlays** [reply]
So it is right that we have to agree to a syntax first?


Maybe we/you can agree on a syntax and let the community implement the feature.

**Reply by jcamiel** [reply]
Syntax is a first step and generally we try to look to other formats to see if there is something already existing (like in HTTP Client from Jetbrains or VSCode Rest Client format). For some "core" features, we prefer to implement it ourselves and have a lot of feedbacks from the community on the design!

**Reply by DenuxPlays** [reply]
> Syntax is a first step and generally we try to look to other formats to see if there is something already existing (like in HTTP Client from Jetbrains or VSCode Rest Client format). For some "core" features, we prefer to implement it ourselves and have a lot of feedbacks from the community on the design! 

Okay understand.
Yeah code reviews etc can be exhausting.
As I said when I find some time I'll comeback with some ideas.

**Reply by DenuxPlays** [reply]
Hey @jcamiel,
I think I'll came up with something.
These are the requirements I wanted to be met (If I missed something please let me know I'll update the syntax accordingly):
- allow multiple imports/dependencies
- allow multiple exported variables
- allow to capture variables from imported file
- the file to import must explicitly export variables that can be latered captured and reused
- Imported variables should be able to be renamed
- allow no variables to be exported (just run another hurl file)

# Basic explanation

## Export

```
/* Request */
[Captures]
export_var1: jsonpath "$.id"

# Explicitly export the token variable
[Exports]
export_var1
```

## Import

```
# Basic import (just execute the hurl file before this one and give me exported variables) 
IMPORT `relative path`
# Rename exported variables
IMPORT `relative path` AS (export_var1 as var1, exported_var2 as var2)
```

# Example

This is an example demonstrating the use of multiple imports, renames and exports.

```hurl
# login.hurl

POST https://api.example.test/login
Content-Type: application/json
{
    "username": "testuser",
    "password": "password123"
}
HTTP/1.1 200
[Captures]
token: jsonpath "$.token"

[Exports]
token
```

```hurl
# create_account.hurl

POST https://api.example.test/accounts
Authorization: Bearer {{token}}
Content-Type: application/json
{
    "name": "New User",
    "email": "newuser@example.com"
}
HTTP/1.1 201
[Captures]
account_id: jsonpath "$.id"

[Exports]
account_id
```

```hurl
# just_a_test.hurl

GET https://api.example.test/some_endpoint
Authorization: Bearer {{token}}
HTTP/1.1 200
```

```hurl
# main_test.hurl

IMPORT login.hurl AS (token AS user_token)

IMPORT create_account.hurl AS (account_id AS new_account_id)

IMPORT just_a_test.hurl 

# Use the captured and renamed variables
POST https://api.example.test/another_request
Authorization: Bearer {{user_token}}
Content-Type: application/json
{
    "account_id": "{{new_account_id}}"
}
HTTP/1.1 200
[Asserts]
jsonpath "$.status" == "success"
```

# Please give me Feedback

It would be great if the core team can review these proposal and give feedback.

**Reply by DenuxPlays** [reply]
One flaw I figured it that you cannot pass data TO the imported file.
Maybe I'll have to think about this too

**Reply by DenuxPlays** [reply]
> One flaw I figured it that you cannot pass data TO the imported file. Maybe I'll have to think about this too

Perhaps something like this could work? (Not sure just sketching atm):
```hurl
IMPORT login.hurl
[Pass]
username = "testuser"
password = "password123"
[Rename]
token AS login_token
user_id AS login_user_id

IMPORT create_account.hurl
[Rename]
account_id AS new_account_id
account_email AS new_account_email
```

**Reply by DenuxPlays** [reply]
Or just do something more familiar with the current syntax:
```hurl
Import login.hurl
[Pass]
username = "testuser"
password = "password123"
[Capture]
login_token: token
login_user_id: user_id

# Import account creation procedure and rename variables
Import create_account.hurl
[Capture]
new_account_id: account_id
new_account_email: account_email
```

**Reply by jcamiel** [reply]
Hi @DenuxPlays thanks for the feedback.

Just to be crystal clear, we valuate good feedbacks / proposal like yours, but, please, do not take too much time into trying to implement it: we don't want people to lost their time!! We prefer implement "core" features ourself and there is very low chance that we integrate a PR on such features.

That said, my two cents on your ideas:

- import / export semantics: I've the impression (maybe wrong) that "export" is used to export variables and "import" is used to include Hurl file. IMO, these are two complementary features that could be addressed separately. The wording lets think that there're related so it can be confusing.
- exporting variables: I like that exporting variables is explicit. What I'm not sure is how to deal with running tests in parallel (it will be the default on the next version). As of 5.0.0 Hurl version, nothing is shared between two Hurl files to facilitate running tests in parallel (think like tests units isolation)
- syntax: if we support including Hurl file (which is not yet stated), this syntax:

```
INCLUDE foo.hurl

GET https://bar.com
```

`INCLUDE foo.hurl` is to similar to a custom request with an custom `INCLUDE method`. Fro instance, this Hurl file is valid:

```hurl
INCLUDE https://foo.com

GET https://bar.com
```

And equivalent to these curl calls:

```shell
$ curl --method INCLUDE https://foo.com
$ curl https://bar.com
```


Once again, thanks for the feedbacks,

**Reply by DenuxPlays** [reply]
Hey @jcamiel,

I just wanted to push that feature a little bit closer to the implementation.
I really want/need this. 
Its understandable that the core team wants to adress this  but you already said that you are not currently working on it.
I am not acutally coding this (unless you want me to).
Its just a concept I wanted to discuss.

I know that parallel tests are default in v5 and this is acutally a bad thing for me.
This will break all of my tests. (Not unfixablee but still)
If this feature would be implemented in v5 I could setup my tests like I initially wanted to without copy-pasting 50 lines of code into **every** hurl file.

I agree that import/export and you include are 2 different features.
The include would be very basic and I guess does not allow to pass variables to it and from it (like the feature I need and others probably want).

The include feature would be very very basic and I guess that this is not really what a lot of people miss in hurl.
And I am not really sure what a Include http method is as I've never heard of it and could not find any documentation about it.

A quick question at the end:
Are you interested to add a feature ike I described (the include feature on streoids)?
Because if not then I think this discussion can be closed and if I think you/we should set some requirements for this feature and open a tracking issue for it.

**Reply by DenuxPlays** [reply]
One thing I forgot:

You said you are not sure how to handle the import feature when running tests in parallel.

My Idea was that we "merge" the imported files into the "parent" file.
So if two "parent" files import the same "child" the "child" there will be 2 "childs" at the same time.
One for each mother file.
Like an instance of the file so that they will be no race onditions etc.

I thnik this would be a pretty good approach.

**Reply by jcamiel** [reply]
> I know that parallel tests are default in v5 and this is acutally a bad thing for me.
This will break all of my tests. (Not unfixablee but still)

You can run your tests in a sequential way like this:

```shell
$ hurl --test --jobs 1 *.hurl
```

This will run your tests in the order they are provided and should work like you're running without parallel option.

**Reply by jcamiel** [reply]
> A quick question at the end:
Are you interested to add a feature ike I described (the include feature on streoids)?

For the moment no, we're focusing development on other features (please don't forget Hurl is a hobby project!!)

**Reply by DenuxPlays** [reply]
Okay
Thats sad :/

Yeah thtas the case with a lot of open-source projects.
Again if I would develop and maintain this feature if wanted. (Maybe not for now but later)

**Comment by jcamiel** [comment]
Just a last answer :) :

> And I am not really sure what a Include http method is as I've never heard of it and could not find any documentation about it.

Like curl, Hurl allows to use any string as a method `curl --method FOO https://bar.com` even if it's not a standard HTTP method (see #967 )
